### PR TITLE
two small fixes

### PR DESCRIPTION
**Pull request by fpvandoorn**
Unrelated remarks:

I've seen this document floating around before, but now I finally read it. It is a very interesting set of conventions, with interesting similarities and differences between the [hanabi.live conventions](https://github.com/Zamiell/hanabi-conventions/blob/master/Reference.md). 

I'm curious about the history of these conventions. Some conventions are exactly the same as on hanabi.live, but with a different name (your *early save* and hanabi.live *5 chop move* or your *layered finesse* and hanabi.live *patch finesse*). Did one group take these conventions from the other group, or were these conventions mostly invented independently?

Also, have you played with the hanabi.live conventions? Do you have an idea how the strength of the conventions compare? I'm in particular interested about the strength of "play all cards until I give a stop clue" in your convention vs hanabi.live's "play leftmost only and keep the other cards in hand".

### CODE REVIEW

**Comment by SashaIr** [comment]
Thanks for the fixes!

So, we started using the _early save_ and the _layered finesse_ independently IIRC, but we indeed took some conventions from Hypen-ated's group (e.g. _chop focus_, which by chance works wonders together with the _cyclic rearrangement_).

I did try many of the hanabi.live conventions, and some of them are great! The _3-bluffs_ and similar are interesting, as for example the _trash finesse_ and other stuff, but we like our _play the complement_ ideas more. Some extra stuff (e.g. _elimination notes_ and similar) are too complicated to use IRL with no note-taking support, so we just ditched them (but we might consider implementing them when playing online). We do use a hat-guessing convention when in a _double discard_ situation: I think it works wonders (except when people miscalculate, but it's only a matter of habit), and the hanabi.live system would greatly benefit from it.

I think overall the hanabi.live system is better, because of the extra layers of complexity you can use when having all the data from the online platform, and because I deliberately avoided adding more stuff because we struggle to get five experts simultaneously on a table (and so I'd rather insist more on mastering the conventions we already have). We might evolve now that we switched to an online platform, however.

Regarding the strength of "play all cards until I give a stop clue" vs "play leftmost only and keep the other cards in hand", that's hard to say. As you read, we play almost exclusively Black (6 suits), which has low pace, high required efficiency, and no ambiguous suits (such as multicolour). I think in that specific framework "play everything" works better, because you really want to play as much as possible instead of discarding, and because the absence of ambiguous suits makes the clues very specific. In the Rainbow (6 suits) variant, "play leftmost only" makes a lot more sense (and in fact we do that, but it's not covered by the document yet), and in other variants like, for example, Pink or Omni, "play all" would be outright crazy. I'm unsure about stuff in between like Dark Rainbow, but I think that one in particular is closer to Black than to Rainbow.

My opinion only though, we didn't really test "play leftmost only" in the Black variant, so I might as well be wrong. If you have any other comment or question, I'm interested in what you have to say!

**Comment by SashaIr** [comment]
I just realised that you don't get notifications for edits!

**Comment by fpvandoorn** [comment]
Thanks for the extra comment, then I indeed saw the notification :)

That is very interesting. Oh yes, the hanabi.live conventions were also evolved from hyphen-ated's conventions.

I saw the hat guessing section, that is a very interesting use for hat guessing.
We tried to also incorporate hat guessing in a different way for very hard variants, if we allow empty clues (empty clues are allowed in some versions of the rules -- out of curiosity, do you know if the Italian version of hanabi allows empty clues?): https://github.com/Zamiell/hanabi-conventions/blob/master/misc/Empty_Clues.md. We usually don't play with this, but it works really well if we do (and make no mistakes).
Being in a double discard situation is an interesting "trigger" for hat-guessing clues (for us giving an empty clue was the trigger). Does it also trigger when someone has the other copy of the discarded card already fully clued in their hand (so it is not a "real" double discard situation)? With the hat guessing clue, sometimes you have to tell people the number of their chop - and then they will chop move it - even though this card was already clued (but not fully clued) in another hand, correct? That is a slight disadvantage, but it's probably still worth the efficiency.

The "play complement" clues and the cyclic rearrangement are very interesting. I have not seen either of those in any other convention system.

Good point that "play leftmost" works better in any multicolor/multinumber variant. With pink we actually do something very similar to "play all" when cluing 1s: if there are unplayable pink cards as part of this 1 clue, we usually fix, and sometimes it works as a delayed multi-card finesse.

The play all convention seems to work best for *Black (6 suits)*, indeed. If you play online at some point, I'm happy to join and learn the conventions better!

**Comment by SashaIr** [comment]
Well, empty clues totally trivialise the hat guessing system: you can just say that an empty clue is a hat guessing clue, and that a regular clue isn't. We don't allow that as it makes the game way too easy. The Italian rules don't specify whether empty clues are allowed or not, and I didn't investigate the matter since, as I said, I think they make the game too easy.

I think that double discard is a very reasonable trigger: it depends only on global information, it doesn't happen that often but it happens often enough for it to be useful (especially towards the beginning of the game), and it's a situation in which you can't really discard but sometimes every regular clue would look ambiguous (and in fact the hanabi.live system relaxes the strength of the meaning of a clue given in a DDA, e.g. allowing _5 Stall_ or similar stuff). Giving it a conventional meaning you get out of the impasse, the clue still gives a lot of information (notice that "0" being an option means that any non-zero number you get is actually a useful card), and in case you have a play or a safe discard you can just pass if the clue is not needed. I'm very satisfied with that choice.

To answer your other questions, the hat guessing clue does trigger even in "fake" DDA situations, because we wanted no risk of misunderstanding, but I guess it can be safely relaxed in case of fully clued cards. Also, it is true that sometimes you have to duplicate cards, but it doesn't happen often and usually one of the involved players realises what's going on very soon and discards their duplicated card. I don't think that has been a major issue in any of our games.

The cyclic rearrangement was born because sometimes it's clear that a player is discarding to prevent the next one from doing so too, and playing IRL you can't really take notes (so we didn't have the "chop move" concept), so in those cases the simplest thing to do was to physically "chop move" your card to another slot. Of course doing so in an arbitrary way passes illegal information to the other players (namely, "I understood that I don't have to discard this"), so we looked for an algorithmic way to achieve the same result. After some trial-and-error, we converged on the current form, which I think is quite nice and works truly wonders with the chop focus convention (before discovering that, we sometimes let playable cards hang on chop hoping that they would shift to slot 1, but that was in general too risky). Now that we can take notes, we could possibly chop move stuff, but I think this convention is good enough to be kept (plus I spent too much time coding it to give it up).

The complement clues look very natural to me, and I'm surprised that no one else uses them. They have been invented when I clued three 4's to someone (and, uhm, gave them some hint to understand what I actually meant), but since then we just formalised the concept and reapplied it also to trash clues towards the end of the game. I have to say, the 3-4 complements work great, we use them at least once per game (and it's often a 4-for-1 clue, sometimes a 4-for-2 if a fix is needed, sometimes a 5-for-1 or more if there are finesses involved) and nobody ever misunderstands them. The trash complement are used a lot less often, but they are also useful from time to time.

We will definitely play online at some point, but when we do we usually have a private voice chat in Italian. Still, if at some point we'll be looking for a 5th player and you're interested we'd be happy to welcome you at our table. I'll tell the others!

**Comment by fpvandoorn** [comment]
> Well, empty clues totally trivialise the hat guessing system

I have heard that from others. I strongly disagree by pointing out that hat guessing already trivializes the game, even with empty clues not allowed.
Here is a description of hat guessing that works perfectly fine without any empty clues needed: https://github.com/Zamiell/hanabi-conventions/blob/master/misc/Hat_Guessing.md
It works great for 4-5 players and reasonably for 3 players.

Also, here are the scores a bot implementation of it achieves, without ever giving empty clues:
https://github.com/fpvandoorn/hanabi#optimized-winrates
Notice that for Black (6 suits) achieves a 56.5% winrate, while an optimized cheating bot (that is allowed to look at its own cards) achieves 72.6% winrate. So room for improvement, but not all that bad.

So IMO the game is "broken" by hat guessing without allowing empty clues, so I don't really see the point of not allowing them.

Thanks for the explanations about the other conventions, they seem neat. I am very interested in the conventions different groups invent. It's also interesting what variant of hanabi is considered the "main" variant. In your group it's Black (6 suits), on BGA it's Multicolor (6 suits). This also influences the conventions: your doc explicitly states that cluing 3s and 4s is pretty useless, which is much more true in Black than in Multicolor.

**Comment by SashaIr** [comment]
Ok, I'll be more precise about why I think that empty clues trivialise the game. There are various reasons, involving both bot players and human ones.

First, sheer numbers. In a _Black (6 Suits)_ game, you are only guaranteed to have 3*(n-1) possible different clues available at any given moment. Several AIs use 4*(n-1), relying on the fact that it's rare to not be able to give the desired clue. If you allow empty clues, then the number skyrockets to 11*(n-1). Considering that the deck is composed of 30 different cards, you can fully clue one card per player and still have room for improvement: you can even do that in three players, provided a tiny bit of optimisation (e.g. you don't really need to distinguish trash cards).

Take for example a 5 player game. After one full round (5 clues), every player has full knowledge about their hand. After that, you only need one clue every four played/discarded cards to keep this status: the required pace is three played cards for each discard, so that's a bit on the edge, but you still have three clues in the bank plus the ones you get for playing 5's. I honestly think that this should achieve the same results as a cheating bot, because the amount of information you get with a clue is high enough to give you full knowledge - and I'm suggesting the trivial strategy of just numbering the cards from 1 to 30, not even doing anything more refined. Seriously, even human players could easily do this if they wanted to.

Second, easiness of computation. With 12-16 choices per player, you need a smart way to give clues (see for example [this](https://github.com/giove91/hanabi) bot from two players in my group, that has a > 63% win rate in _Black (6 suits)_ with an average score of 29.38) and about which card. With 44 choices, you can straight away give a card to each player without even thinking about what you're doing. Then it's also trivial to decide whether a card is known or not, because when you get info about it, you get full info, and there's no need to decide a priori which one is the card with the least information on (which is part of what the linked bot does). As I said, with empty clues, even human players can achieve cheating-bot level results.

Third, the possibility to choose a framework. With human players and empty clues allowed can just pick a 12-choices system, and then, at any given moment, you can decide whether to give a regular non-empty clue, with all the implications of it (finesses etc.), or a hat-guessing empty clue. No risk of misunderstanding, no risk of giving useless information, nothing. You can just pick every time the one that's more convenient, _and that won't even be the best strategy_.

Summing up, I believe that a bot that uses empty clues well can easily perform as well as the cheating bot, and that the algorithm is easy enough to be used by humans, and strong enough to work in more complex variants (e.g. even very hard variants like _Dark Rainbow (5 Suits)_ or _Gray (5 Suits)_ allow for 9*(n-1) choices, more than one for each different card in the deck even in 4 player games). Without empty clues, this is just not possible, as the range of possible choices for a clue narrows too much: maybe the hat-guessing system trivialises the game on its own, but empty clues trivialise even the hat-guessing system itself.

**Comment by SashaIr** [comment]
Oh, regarding the last part, we consider "Black" the main variant for the very stupid reason that the first Hanabi deck we bought only listed that as a variant (and indeed it featured only 55 cards in total). I think the reason is that an industrial printing sheet has only space for 57 cards in the "8x8 square" pocket format, and so it's more convenient for the producers to print 55 cards instead of 60. The rectangular format does not have this issue, so they provide 60 cards instead, but we were so used to _Black (6 Suits)_ that even when we upgraded to Hanabi Extra (the nice one with big cards and card holders) we stuck to that variant.

**Comment by fpvandoorn** [comment]
I wrote a reaction earlier, but apparently I forgot to send it.

I agree with all your points of empty clues making hat guessing easier. I do have a couple of points:
* I think that even with empty clues it's not easy to achieve the same results as a cheating bot (and to be honest, it's not even easy to write a cheating bot that plays perfectly; I wouldn't be surprised if someone comes with a cheating bot that does more than 1 percentage point better in black)
* We are speculating about three bots: one that doesn't give empty clues, one that does give empty clues and a cheating bot. I wouldn't be surprised if the former two are closer in performance than the latter two, when programmed optimally.

But yes, I agree with your main point that hat guessing is a lot easier with empty clues.
